File: timeseries/timesense/keys.py
"""API key utilities with macOS Keychain fallback."""

import os
import subprocess
import logging

logger = logging.getLogger(__name__)


def get_api_key(key_name: str = "OPEN_ROUTER_KEY") -> str:
    """Get API key from environment or macOS Keychain.

    Args:
        key_name: Key name to look up (used for both env var and keychain service)

    Returns:
        The API key string

    Raises:
        ValueError: If key not found in environment or Keychain
    """
    key = os.getenv(key_name)
    if key:
        return key

    # Fall back to macOS Keychain (same key name used as service)
    try:
        username = os.getenv("USER")
        result = subprocess.run(
            ["security", "find-generic-password", "-s", key_name, "-a", username, "-w"],
            capture_output=True,
            text=True,
            check=True
        )
        return result.stdout.strip()
    except subprocess.CalledProcessError as e:
        logger.error(f"Failed to get API key from Keychain: {e}")
        raise ValueError(f"Cannot find {key_name} in either environment or keychain")

File: src/clajent/keys.clj
(ns clajent.keys
  (:require [clojure.java.shell :as sh]
            [clojure.string :as s])
  )

; security add-generic-password -s OPENAI_API_KEY -a "$(whoami)" -w "$(pbpaste)"
; security add-generic-password -s OPENAI_API_KEY -a "$(whoami)" -w "$(pbpaste)" -U
; security find-generic-password -s OPENAI_API_KEY -a "$(whoami)" -w | pbcopy

(defn get [^String k]
  (or (:out (System/getenv k))
      (-> (sh/sh "security" "find-generic-password" "-s" k "-a" (System/getProperty "user.name") "-w")
          :out
          s/trim)
      (throw (IllegalArgumentException. (str "Cannot find " k " in either environment or keychain")))
      ))
